### DISCUSSION TITLE
Add Example: With DI

### DISCUSSION THREAD

**Discussion by one-aalam** [livebud/bud#72] [Ideas]
Bud's impressive and I understand it's just `0.1`, so it would be highly ambitious(and very, very early) to expect a very good documentation, blog and example ecosystem. But, just to give it a consideration as a very low priority task - _will it be possible for you to consider adding an example with DI(Env, DB, etc.)_ ? I'm not a seasoned Go dev, and I think seeing a good example with DI can help anybody like me, even the developers that Bud might inevitably attract due to its affinity towards modern FE frameworks.

**Comment by matthewmueller** [comment]
Thanks for opening this issue @one-aalam! I'll definitely need to add more examples. 

Did you see: https://www.notion.so/Hey-Bud-4d81622cc49942f9917c5033e5205c69#976e4b0666f34d5eae6ccbce93111025? Could I improve the example in there?

One more example I can share now: In [the demo video](https://youtu.be/LoypcRqn-xA?t=245), I show, but don't explicitly mention the dependency injection.

That `*hackernews.Client` is automatically wired up. The function it calls is here: https://github.com/matthewmueller/hackernews/blob/6eef99e306e59c859d5e14e8ca655acb31cd5fe7/client.go#L23-L25

Basically DI looks for a way to provide a `*hackernews.Client` to the `Controller`, it finds that if it calls the `New()` function it will be able to provide it, so it generates the code to call the `New()` function. 

_(FYI I converted this issue to a discussion since it's not a bug)_

**Comment by one-aalam** [comment]
Thanks @matthewmueller [This](https://www.notion.so/Hey-Bud-4d81622cc49942f9917c5033e5205c69#976e4b0666f34d5eae6ccbce93111025) should do for now.

I was looking for an example with DB, env, etc. that shows how something like a DB connection is established/released(for different environments), provided in `main`, injected in modules and later used for querying and updates(like a full CRUD example with DB)

**Reply by matthewmueller** [reply]
Ah, so you can do this manually or wait for [v0.3](https://github.com/livebud/bud/discussions/21) which will ship with environment support.

To do it manually, you can use https://github.com/caarlos0/env and do the following (untested):

1. Create an `env/env.go` file:

```go
package env

import "github.com/caarlos0/env"

// Load the environment
func Load() (*Env, error) {
  var e Env
  if err := env.Parse(&env); err != nil {
    return err
  }
  return &e, nil
}


type Env struct {
  DatabaseURL string `env:DATABASE_URL`
}
```

Then in a `db/db.go` file:

```go
package db

import "app.com/env"
import "github.com/jackc/pgx"

func Dial(env *env.Env) (*Client, error) {
  return pgx.Dial(env.DatabaseURL)
}

type Client = pgx.Client
```

Finally in your `controller/controller.go`:

```go
package controller

import "app.com/db"

type Controller struct {
  DB *db.Client
}

type User struct {
  Name string
}

func (c *Controller) Index() ([]*User, error) {
  rows, err := c.DB.Query("select * from users")
  if err != nil {
    return nil, err
  }
  // TODO: turn rows into users
  return users, nil
}
```

With that you can now switch out your development database with your production database using:

```sh
DATABASE_URL="..." bud run
```

In v0.3, Bud will ship with 4 default environments and a bit more structure around how to do environment management. This example is the basis for the support though.

Please give it a try and let me know if you run into anything or have anymore questions!

**Reply by one-aalam** [reply]
Thank you, @matthewmueller that's all! :-)

Looking forward to 0.3 which will have the env support baked in, but this is more than enough for me for now.
